### DISCUSSION TITLE
Development across the Fediverse

### DISCUSSION THREAD

**Discussion by CaptainCalliope** [owncast/owncast#3093] [General Owncast discussion]
I'm not sure where to put this, so here is as good a place as any to start. This is going to be a bit of an unstructured brain dump. Apologies if overwhelming.

I think there should be a place to track opportunities to do Owncast Fediverse development outside the Owncast codebase.

Some questions on my mind:
* What would it look like for other Fediverse apps to add features that benefit and/or incorporate Owncast directly? 
* For each third party app and platform, what are changes or additions they can make to their UIs? 
* What can be added that takes advantage of the current API? 
* Are there features we'd like added to third party platforms and apps that could be made easier by new additions to the API? 
* Could we identify things missing from the ActivityPub spec that could grow the livestream ecosystem?
* Could sdks be created to enable third parties to easily add Owncast specific functionality to their apps and platforms, including streaming to Owncast from them?

Some feature ideas I have for outside apps, in no particular order:
* When I see a post/update from an owncast server in my Mastodon/Calckey/Pixelfed/etc newsfeed, I want to be able to click on it and play it right there.
* I want to be able to stream from whatever fediverse phone client I'm using. I'd prefer third-party Fediverse clients to be Owncast enabled before an official Owncast client.
* Specifically, I want Owncast deeply integrated with the Pixelfed experience for both viewers and creators.
* From a viewer perspective, I want a first-class experience from whatever phone app I'm in. If I'm logged into my Mastodon client and click on a livestream, I want a playing livestream to either stick around as an overlay or stick to the top part of my phone as I scroll my feed. If I click on it, I want the stream + chat experience. If I hit the back button, I'm back to my newsfeed. Very much a Twitch-like mobile experience, except the home screen is the newsfeed instead of other livestreams.
* If people are viewing from whatever app, they don't have to authenticate, they just have to hit a "join chat" button. (Join chat anonymously could also be an option)
* As a creator, I want my profile page to be viewed as clearly a livestream page rather than just another fediverse profile inside of all fediverse apps, but with whatever layout makes the most sense for each app. My buttons and tabbed content should be available programmatically so that people can take advantage of whatever interactivity I've set up between my viewers and my livestream.
* Etc, etc etc.
* Etc.

My big question is, how can Owncast become a catalyst for livestreaming to become a first-class feature/experience across the fediverse, and how can that be tracked?

I feel like tracking feature requests for other apps and platforms in a single place can inform what new features and priorities for the Owncast codebase. I also feel like tracking this stuff in one place could inspire third party platform and client devs to want to work on Owncast related features. I also also also think this could inspire alternatives to Owncast to spring up to plug into these experiences once they reach a level of ubiquity. How cool would it be if non-Owncast streams started showing up in Owncast directories?

And could a side effect of this include federated events and calendaring to finally be boostrapped as a core part of the overall fediversal experience?

Anyways, I know pretty much all the work right now is being done by Gabe, but again part of my thinking here is documenting this stuff could be enough to invite other devs to chew on Owncast as a platform.

A side note: Calckey has [a plugin and widget system](https://box464.com/posts/calckey-plugins/). Is there something extreeeeeemly lowhanging that could be prototyped for it? Perhaps something similar to [this Owncast browser extension]( https://addons.mozilla.org/en-US/firefox/addon/owncast-extension/) showing up in the Calckey sidebar? 🤔

**Comment by gabek** [comment]
I'll try to address things best as I can.

I'm not sure how to best put this into words, but I'll try. The Fediverse is 110% supportive of Owncast and really feels positively about it. However, the people on the Fediverse, in general, don't use Owncast. You can see #3088 for specific data around this. I've been watching how people run **public** streams for a while and these are the scenarios I see:

1. Individuals who think Owncast is cool will set it up, and stream once or twice, realize they don't really have a reason to use it, and never stream again.
2. Individuals who are super excited about starting a stream start streaming but realize they don't have an audience and stop streaming.
3. Organizations, events, and streams that don't rely on a human streaming, or viewer interaction run regularly.

If you look at the directory the streams that stick around are the 24/7 tv or radio streams, this is because Owncast is solving a real problem for them. They want to stream to their **existing** audience and Owncast allows them to do that. So they run it for years.

The rest of Owncast use are **private** streams that specifically choose to not have a Fediverse (or directory) presence. Due to them being private I can only guess how many these might be or what style of streams they are. But I do know wedding venues, music venues, conferences, one-off events. These are the people who use currently use Owncast and rely on it. None of these are on the Fediverse or want to be public.

I keep track of these things, and the number of public streams is mostly flat over time. That means for every new person who tries out Owncast to start a stream, another person stopped. However, the number of downloads of Owncast continues to grow. It's being used, but we don't see it. They're not replacing Twitch, they're building something new.

I only bring this up because unfortunately the people who use Owncast the most are not the people on the Fediverse. The people on the Fediverse **support** Owncast, but they're not the ones using it regularly. People call it a "Twitch alternative", and while you can use it that way that's not how people are most successful with it. So this means focusing specifically on the "Fediverse individual" use case I don't believe is the best use of effort at this time.

I clearly want to grow the Fediverse functionality, however. But this means building features that work with or without Fediverse functionality enabled, but if it is enabled, allows for greater reach. Such as schedules or clips. These are useful if you're public, or private, or not on the Fediverse.

> When I see a post/update from an owncast server in my Mastodon/Calckey/Pixelfed/etc newsfeed, I want to be able to click on it and play it right there.

With Mastodon on the web this is already possible, but I'm pretty sure Mastodon web is the only interface that supports inline rendering of video. It's up to every interface to decide how to render links. Owncast supports the Opengraph standard for rendering metadata, but everyone treats it differently.

> I want to be able to stream from whatever fediverse phone client I'm using. I'd prefer third-party Fediverse clients to be Owncast enabled before an official Owncast client.

If you use a client to interact with Mastodon or Pleroma it doesn't make a lot of sense to have live stream functionality in that client, because Mastodon or Pleroma or Misskey doesn't do livestreaming. A native Owncast application would make more sense, but that's not in the cards.

> For each third party app and platform, what are changes or additions they can make to their UIs?

Not even UIs, it would be great if other platforms didn't completely break. Right now Mastodon, for example, completely throws away an ActivityPub activity if it has more than one object is in it. Meaning I can't send a text post and an Event (for example) at the same time. This is filed in https://github.com/mastodon/mastodon/issues/19111 and they don't care. It's the mindset that "we don't deal with Events, so this isn't our problem". But because Mastodon is the biggest project now it's my problem because I do want to support events.

It would be great if an Event came through and in UIs you could add to calendar, or "Join stream" or something actionable. When it comes to the Fediverse it's all about trying to encourage more people to view more streams and help people build up the audience they don't yet have.

> Specifically, I want Owncast deeply integrated with the Pixelfed experience for both viewers and creators.

I tried to get Owncast stream previews federating with Pixelfed at one point, and it didn't work. I didn't know why, and I knew it wouldn't move the needle, so I just moved on. I don't use Pixelfed myself so I'm a bit detached from that world.

> From a viewer perspective, I want a first-class experience from whatever phone app I'm in.

Remember that when talking about standards these platforms don't know that this is an "Owncast stream", it knows there's some content that can be embedded inline. It doesn't know that there's chat, it just knows to draw an iframe on the web page. To support Owncast chat natively in 3rd party UIs, while technically possible, would be mostly impossible with all the moving parts and different parties involved.

> As a creator, I want my profile page to be viewed as clearly a livestream page rather than just another fediverse profile

While technically ActivityPub could be expanded to support additional metadata describing a style of ActivityPub actor, we've kind of standardized on person, bot and service based on what ActivityPub already offers us. Adding something specific like "livestream" would never go anywhere, and nobody would put the effort into making custom interfaces for only a few accounts.

All this is to say, Owncast + Fediverse right now is too small for it to be the focus for me, much less the focus for anybody else. Other platforms **really** don't care about Owncast. And sure, all this could change tomorrow if individuals decided they wanted to live stream and really work on building an audience. But right now people on the Fediverse are not actively, seriously, using Owncast. And many who do, want to keep their streams private and off the Fediverse.

If you have ideas on how to improve this, so the Fediverse can become a priority, and Owncast is big enough for other platforms to care, I'd love to hear about any thoughts you might have!

Edit:

I realized I might be a bit dismissive of those individuals who **do** stream using Owncast, and **have** put effort into building an audience and continue streaming. These people are fucking amazing and I owe them so much more than they could ever realize. But these people are the minority. We're talking less than a handful of people fall into this category.

Edit 2:

And so it doesn't sound like I'm doing nothing, I'd really like to see presence indication standardized across the Fediverse. For Owncast it would mean that it shares that you're currently live, and render that across other platforms. This potential standard was proposed at https://github.com/gabek/fedidocs/pull/10.


Edit 3:

I wanted to show my data. Top line is the number of Owncast servers on the Fediverse. The middle line is the number of Owncast servers listed in the directory (meaning they've actually streamed in the past month), huge discrepancy. Bottom line is the number of streams currently live (as known by the directory).

![image](https://github.com/owncast/owncast/assets/414923/9619a946-d705-45db-b945-59137708c1c7)

**Reply by tsmethurst** [reply]
Sorry it's off topic, but I just wanna drop in to say that I very much appreciate your work and your attitude to the work @gabek ❤️

**Reply by gabek** [reply]
Thank you so much, Tobi! That's so kind of you, it means a lot!

**Comment by jprjr** [comment]
Re: playing the video while browsing apps like Calckey, Mastodon, etc.

Just thought I'd share that Calckey supports playing the livestream from Owncast instance, I'm assuming thanks to the OpenGraph tags that specify how to embed the stream. This works on mobile as well as desktop, my Calckey instance is still on the latest stable release (13.1.4.1, released about 4 months ago) so it's been available for at least as long as that, though probably longer.

I do have to specifically tap/click the "play" icon that appears (anywhere else takes me to the page). I can then start playing the video, and click/tap the picture-in-picture icon and browse my home feed while the stream continues to play.

If I change to say, the local feed, or a list, tap a post, or whatever, it goes away.

I do notice the `og:video` url is to the video embed url. I wonder if linking to the HLS playlist would trigger/enable different behavior?

(long thoughts on HLS versus embedded page incoming):

I'm guessing that the main benefit of linking a webpage is: as long as the client can embed an iframe, you're good to go. Pretty sure all Calckey is doing is slapping a play icon over the preview URL, tapping the play icon loads the embed, then it just lets the embed do its thing. Since the embed is handling things like picture-in-picture, the Calckey client must not be aware that picture-in-picture is active, probably not even aware it's playing a video, etc. It's just hosting in iframe. So navigating away from the iframe makes everything go away.

If instead, the `og:video` linked to the HLS url, then the client could do things like, use its own video controls, be aware of events like the video is playing, picture-in-picture is happening, etc - it generally becomes more aware of what is going on, but the drawback is it needs to specifically support videos now. Which they probably do, but if the `og:video` is a URL to a video file / HLS playlist, they're probably just going to put it in a `<video>` element and hope the browser can play it.

One thing that's a little frustrating about HLS is how browsers don't natively support it (except Safari), and apps like Mastodon, Calckey, etc may not be interesting in bundling an HLS library into the web clients. Native apps may have a better time, since they'll probably fall back to a native-ish video player on the OS (or like an Android, you'll probably bundle a library like ExoPlayer anyway), and said player may already support HLS so the app doesn't really have to do anything. So all that to say, linking directly to the HLS playlist probably won't work too well?

So, just some thoughts there.

**Comment by gabek** [comment]
> I wonder if linking to the HLS playlist would trigger/enable different behavior?

> One thing that's a little frustrating about HLS is how browsers don't natively support it (except Safari)

Yeah exactly, most browsers don't support HLS, Chrome-based browsers don't, so it's kind of a deal breaker in that respect. Expecting any of the other projects to build support in I'm sure is a non-starter.

Plus it's nice to have the full Owncast player experience for consistency and metrics (performance, viewer count, branding, etc). Also the offline state, for after the stream ends. It's really nice to be able to have a custom offline message and not just the "Stream is offline" clip.
